### DISCUSSION TITLE
Nested component is re-initialised when LiveView Assigns is updated

### DISCUSSION THREAD

**Discussion by mitkins** [alpinejs/alpine#1404] [7. Bugs (archived)]
I wasn't sure if I should file this as a bug or even if this belongs in the Alpine repo - it's an Alpine behaviour that's caused by a LiveView update. Apologies if this doesn't belong here 😅. Here goes...

I'm currently writing an emoji lookup feature for my textarea. When you type the `:` character, you'll see a lookup menu 😸. As you type each character, it'll filter the list. On my search page (which has one Alpine component), this works fine. On my main data entry page, the lookup disappears and works intermittently:

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/1645205/100806890-62558a80-3485-11eb-88f1-bbb8fb0a46d1.gif)

I've figured out that this is a combination of when there are nested Alpine components and Live View updates the page (after an Assign has changed). So I created a small test case to reproduce the issue. Here's the html from my LiveView:

```html
<div x-data="{}" x-init="console.log('Parent init')">
  <p>Live View Bob: <span><%= @bob %></span></p>

  <div x-data="{ bob: '' }" x-init="console.log('Child init');">
    <p>Alpine Bob: <span x-text="bob" /></p>

    <%= f = form_for @changeset, "#", [phx_change: "change"] %>
      <%= textarea( f, :text, x_model: "bob" ) %>
    </form>
  </div>
</div>
```

And here's the code:

```elixir
defmodule ZoinksWeb.TestLive do
  use ZoinksWeb, :live_view
  alias Zoinks.Search.Criteria

  def mount( _params, _session, socket ) do
    socket = socket
    |> assign( :bob, "bob")
    |> assign( :changeset, Criteria.changeset(%Criteria{}, %{}) )

    { :ok, socket }
  end

  @impl true
  def handle_event("change", %{"criteria" => %{"text" => text}}, socket) do
    socket = assign( socket, :bob, text )

    { :noreply, socket }
  end
end
```

Here's the output after I've typed something into the textarea:

![image](https://user-images.githubusercontent.com/1645205/100807439-6e8e1780-3486-11eb-9afc-a95994872e6f.png)

The Live View binding works fine, but the Alpine binding is empty! Looking at the javsacript console, it seems that the child Alpine component is being reset and `x-init` is being called constantly:

![image](https://user-images.githubusercontent.com/1645205/100807624-c88edd00-3486-11eb-868f-df2674824357.png)

There are 2 ways I can make the code work again. Either I remove the parent Alpine component or I comment out this line in the `handle_event` function:

```
socket = assign( socket, :bob, text )
```

So I'm wondering, is this intended behaviour? Is this just the mechanics of using Live View and Alpine together? That is, is the Alpine child component reset unavoidable?

I think this may be related to https://github.com/alpinejs/alpine/issues/420. I tried forking Alpine so I could replace the `clone` function (as suggested). But it didn't seem to help. It's the first time I've included a GitHub repo using `npm` - so I could have done something wrong.

My other thought is that maybe I could work around the issue by implementing state management with [Spruce](https://github.com/ryangjchandler/spruce)?

Any thoughts?

**Comment by KevinBatdorf** [comment]
Whenever any element with `x-data` is inserted on the page, Alpine will initialize it as a component. You can use `clone` to copy the current state of the  _just-removed_ component to the _about-to-be-inserted_ component and Alpine will clone the state, but not run all the init stuff. 

I don't know how LiveView works, but maybe there's a hook in there? Maybe https://github.com/phoenixframework/phoenix_live_view/issues/809#issuecomment-637671803

I don't think that solves your flashing stuff, unless it's as a result of code being run in `x-init`

**Comment by mitkins** [comment]
Ah yes, I should have mentioned that I have that code snippet setup in my `app.js`:

```javascript
let liveSocket = new LiveSocket("/live", Socket, {
  params: { _csrf_token: csrfToken },
  hooks: Hooks,
  dom: {
    onBeforeElUpdated(from, to) {
      if (from.__x) {
        window.Alpine.clone(from.__x, to)
      }
    }
  }
})
```

However, it's a little bit more than a flash - `bob` in the child Alpine component is constantly being reset to an empty string - in spite of the `x-model` attribute. Though the contents of the textarea isn't affected.

My understanding is that Live View should only re-render the `<%= @bob %>` portion of the html (when the Assigns change). So I'm not sure why the child component is being re-initialised. Though I don't understand the internals of what would actually happen to the html in this scenario.

Also, I removed the `x-init` calls, just in case. I still get the same result

**Comment by mitkins** [comment]
The more I think about it, perhaps this is a Live View issue? Ah but then why doesn't it happen when there's only 1 Alpine component?

**Comment by KevinBatdorf** [comment]
There's also this issue: https://github.com/alpinejs/alpine/issues/756

Do you have events on the child component? Can you console.log those to see if those are being called twice after diff? That could explain the flashing.

**Comment by mitkins** [comment]
In the case of the emoji code, yes I do. I added a `console.log` for each of the events. Here is a snippet from my real code:

```html
      <div x-data="autocomplete()"
        x-init="$watch( 'text', value => { console.log('watch called!'); checkAutocomplete(); } ); text = '<%= input_value f, :comment %>'; console.log('yes?')"
        x-on:keydown.arrow-up="console.log('keydown.arrow-up called!'); move( $event, -1 )"
        x-on:keydown.arrow-down="console.log('keydown.arrow-down called!'); move( $event, 1 )"
        x-on:keydown.enter="console.log('keydown.enter called!'); keydown_enter($event)"
        x-on:click.away="console.log('click.away called!'); hasFocus = false">
        <%= textarea( f, :comment,
          class: "form-input comment block -mt-1",
          autocapitalize: "sentence",
          autofocus: (@live_action == :index),
          x_model: "text",
          "x-on:focus": "console.log('focus called!'); hasFocus = true",
          "x-on:change": "console.log('change called!'); hasFocus = true",
         ) %>

         <div class="absolute z-10">
           <%= if Enum.any?(@suggestions) do %>
             <ul x-ref="suggestions" x-show="autocomplete && hasFocus" class="border rounded bg-white mt-1 list-none max-h-64 overflow-auto">
               <%= for {emoji, index} <- Enum.with_index(@suggestions) do %>
                 <li id="item-<%= emoji.short_name %>"
                   x-ref="item-<%= index %>"
                   x-on:click="select( $event, $refs['item-<%= index %>'] )"
                   data-short-name="<%= emoji.short_name %>"
                   class="px-2 py-1 hover:bg-blue-500 hover:text-gray-100"
                   :class="{
                     'bg-lightBlue-200': index === <%= index %>,
                     'text-gray-600': index === <%= index %>
                   }">
                   <%= ZoinksWeb.EmojiView.render("emoji.html", %{short_name: emoji.short_name}) %>
                   <span class="ml-1">:<%= emoji.short_name %>:</span>
                 </li>
               <% end %>
             </ul>
           <% end %>
         </div>
      </div>
```

As you can see, there's also a `$watch` call in there. None of the events trigger twice. I did have an issue where old events were firing if you clicked on a suggestion (based on position in the list), but I resolved this by giving each item an `id` that was tied to `emoji.short_name` (instead of `index`).

**Comment by mitkins** [comment]
I have an update. I added some debugging to the Live View code:

```javascript
    onBeforeElUpdated(from, to) {
      if (from.__x) {
        console.log(">>>", from.getAttribute("x-data"), (!to.__x ? 'Needs to clone' : 'All good') )

        window.Alpine.clone(from.__x, to)

        console.log("<<<", from.getAttribute("x-data"), (!to.__x ? 'Needs to clone' : 'All good') )
      }
    }
```

Here's the output in the console on initial page load:

```
Parent init
Child init
>>> {} Needs to clone
Child init
<<< {} All good
>>> { bob: '' } All good
<<< { bob: '' } All good
```

As expected, the elements are re-wired/cloned on second mount. Here's what I get after typing a character:

```
>>> {} Needs to clone
Child init
<<< {} All good
>>> { bob: '' } All good
<<< { bob: '' } All good
```

Because the parent is cloned before the child, the child `x-init` is being called in `walkAndSkipNestedComponents`:

https://github.com/alpinejs/alpine/blob/6ad16ab7285d76ae8eb043c9f14337fa25a85564/src/component.js#L199-L209

I see code in there that prevents the walk for child components of an `x-for` or an `x-if`. 

Maybe this is a special case too?

**Comment by KevinBatdorf** [comment]
That makes sense actually (It's being initialized on line `187` though, I believe). It might make sense for Alpine to track whether a component is being cloned then skip a `x-init` for its child elements. 

Have you tried cloning the child first though?

```js
onBeforeElUpdated(from, to) {
  if (from.__x) {
    console.log(">>>", from.getAttribute("x-data"), (!to.__x ? 'Needs to clone' : 'All good') )

    window.Alpine.clone(from.querySelector('[x-data]').__x, to.querySelector('[x-data]'))
    window.Alpine.clone(from.__x, to)

    console.log("<<<", from.getAttribute("x-data"), (!to.__x ? 'Needs to clone' : 'All good') )
  }
}
```

I'm not entirely sure that will work but maybe worth trying?

**Comment by mitkins** [comment]
I didn't think of that! Slightly different version:

```javascript
    onBeforeElUpdated(from, to) {
      if (from.__x) {
        const child_from = from.querySelector('[x-data]')
        const child_to = to.querySelector('[x-data]')

        if ( child_from && child_to ) window.Alpine.clone( child_from.__x, child_to )

        window.Alpine.clone(from.__x, to)
      }
    }
```

So yeah, that's an okay temporary workaround. But it'd be nice if Alpine could keep track of this for me, assuming it's the right thing to do for Alpine.

Or if it were possible, reverse walk up the tree?

**Comment by KevinBatdorf** [comment]
Might be able to pass in the callback as a parameter on `walkAndSkipNestedComponents` so that if the walker encounters a nested component it can handle it differently. I can make a PR this weekend and see how it goes.

**Comment by KevinBatdorf** [comment]
Was thinking... the diff could have new components being added that require `x-init` to run, so it would be impossible to know what to init and what to ignore, I think. 

If you don't need to preserve the child component state, you could just to remove all child `x-init` attributes.

Maybe like this: 
```js
onBeforeElUpdated(from, to) {
    if (from.__x) {
      Array.from(from.querySelectorAll('[x-init]')).forEach(element => element.removeAttribute('x-init'))
      // ...
    }
})
```

...then it won't init any more. 

You could even remove them on the component itself early. It's not needed after it runs once.

```html
<div x-data x-init="doSomething();$el.removeAttribute('x-init')">
```

But if you need to preserve the child component state, you will need to traverse and clone manually, or just _don't use nested child components here_ (if that's an option).

Maybe someone else will have a better answer though. Livewire could even have something built in for this, and I don't have much experience with how Livewire works.

**Comment by mitkins** [comment]
Yes, I came to a similar conclusion about picking up new Alpine Components. I've been trying to understand the code in Alpine so I can think of a solution. These are my thoughts so far:

* You want Alpine to do whatever it does now on the first page load
* This is an assumption, but I think the goal of Live View (on second load) is to walk the DOM and re-create any Alpine objects that were attached to old elements
* So my thought is that in this case Live View doesn't want to do a "deep clone" - because it's going to walk the DOM anyway. So perhaps I could just simply skip DOM walking in this case? 

Potentially, as Live View walks the DOM, the child Component would be treated as a root node - `componentToClone` is passed up to the `Component` and the existing code in Alpine would skip the call to `x-init`.

Of course I tried this and it didn't work (because it's not that simple)! I need to learn more about Alpine to figure out what I'm missing. But that may be besides the point. Do you think there is merit to the idea of a "shallow clone"?

My other idea is that I could create a "Live View init" - that would be ignored during the Live View clone. This is where I could put my `<%= input_value f, :comment %>` and I could let Alpine do whatever it wants with `x-init`. However, I'm not sure whether this is a good idea either.

Regardless, I'll try your suggestions too. I'll let you know how that goes. 

At this stage, it seems that I may be better off with an alternative that doesn't involve nested Components!

**Comment by alvivi** [comment]
I was able to workaround this issue for a complex scenario using this:
```javascript
function cloneChildren(fromChildren, to) {
  for (let i = 0; i < fromChildren.length; i++) {
    const fromChild = fromChildren[i];
    cloneChildren(fromChild.children, to);

    if (fromChild.__x && fromChild.id) {
      const toChild = to.querySelector(`#${fromChild.id}`);
      if (toChild && !toChild.__x) {
        window.Alpine.clone(fromChild.__x, toChild);
      }
    }
  }
}

let liveSocket = new LiveSocket("/live", Socket, {
  dom: {
    onBeforeElUpdated(from, to) {
      if (from.__x && !to.__x) {
        cloneChildren(from.children, to);
        window.Alpine.clone(from.__x, to);
      }
    },
  }.
  ...
});
```
This requires that children must have id. Not ideal, but maybe is useful for somebody :shrug:

**Comment by rubas** [comment]
@alvivi 

Here is a version, that does not need the ids.

```javascript
const cloneAlpineJSData: (from: HTMLElement, to: HTMLElement) => void = (from, to) => {
  if (!window.Alpine || !from || !to) return

  for (let index = 0; index < to.children.length; index++) {
    const from2 = from.children[index]
    const to2 = to.children[index]

    if (from2 instanceof HTMLElement && to2 instanceof HTMLElement) {
      cloneAlpineJSData(from2, to2)
    }
  }

  if (from._x_dataStack) window.Alpine.clone(from, to)

  // Set `x-persist="type,style.height,style.minHeight"` to prevent these attributes from being overrided by an LV update
  persistAttributes(from, to, from.getAttribute('x-persist'))
}

const liveSocket = new LiveSocket('/live', Socket, {
    hooks,
    params: { _csrf_token: csrfToken },
    dom: {
      onBeforeElUpdated(from, to) {
        cloneAlpineJSData(from, to)

        return true
      }
    }
  })
```
